### ISSUE TITLE
error with BAM file coming from minimap2

### ISSUE THREAD

**Issue by lpantano** [wdecoster/NanoPlot#177]
Hi!

thank you for developing this tools, it is quite good. Being new to nanopore is hard :)

I mapped reads with minimap2 and after sorting the file I got this error with version 1.29.0. 

Not sure if it because the bam is malformatted or something is missing due to the nature of my data. This is whole genome sequencing (1x) of pig samples. 

Thanks!

```
/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/kernels.py:125: RuntimeWarning:

divide by zero encountered in true_divide

/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/kernels.py:125: RuntimeWarning:

invalid value encountered in true_divide

/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/_kernel_base.py:514: RuntimeWarning:

invalid value encountered in true_divide

/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/matplotlib/contour.py:1520: UserWarning:

Warning: converting a masked element to nan.

/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/matplotlib/contour.py:1521: UserWarning:

Warning: converting a masked element to nan.

/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/matplotlib/contour.py:1169: RuntimeWarning:

invalid value encountered in less

/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/matplotlib/contour.py:1171: RuntimeWarning:

invalid value encountered in greater

If you read this then NanoPlot 1.29.0 has crashed :-(
Please try updating NanoPlot and see if that helps...

If not, please report this issue at https://github.com/wdecoster/NanoPlot/issues
If you could include the log file that would be really helpful.
Thanks!



Traceback (most recent call last):
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/kde.py", line 451, in kdensityfft
    bw = float(bw)
ValueError: could not convert string to float: 'scott'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/bin/NanoPlot", line 10, in <module>
    sys.exit(main())
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/nanoplot/NanoPlot.py", line 96, in main
    plots = make_plots(datadf, settings)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/nanoplot/NanoPlot.py", line 236, in make_plots
    plot_settings=plot_settings)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/nanoplotter/nanoplotter_main.py", line 204, in scatter
    height=10)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/seaborn/axisgrid.py", line 2325, in jointplot
    grid.plot_marginals(kdeplot, **marginal_kws)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/seaborn/axisgrid.py", line 1810, in plot_marginals
    func(self.x, **kwargs)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/seaborn/distributions.py", line 705, in kdeplot
    cumulative=cumulative, **kwargs)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/seaborn/distributions.py", line 295, in _univariate_kdeplot
    cumulative=cumulative)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/seaborn/distributions.py", line 367, in _statsmodels_univariate_kde
    kde.fit(kernel, bw, fft, gridsize=gridsize, cut=cut, clip=clip)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/kde.py", line 140, in fit
    clip=clip, cut=cut)
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/kde.py", line 453, in kdensityfft
    bw = bandwidths.select_bandwidth(X, bw, kern) # will cross-val fit this pattern?
  File "/opt/tools/conda/envs/WF-NF-nanopore-1.0dev/lib/python3.7/site-packages/statsmodels/nonparametric/bandwidths.py", line 174, in select_bandwidth
    raise RuntimeError(err)
RuntimeError: Selected KDE bandwidth is 0. Cannot estimate density.
```

**Comment by lpantano** [comment]
Sure, I can do that, I will re-generate it, since I was playing with some
data only. Thanks for looking into this!

**Comment by wdecoster** [comment]
Great, you should be able to find my emailadress in my profile

**Comment by weir12** [comment]
Hi:
I encountered the same problem when Executed the following command.
```
NanoPlot -t {threads} --bam {input.BAM} --raw -o {output.DIR} -p {params.sample}_ --N50 --title {params.sample} --downsample 100000
```
bam file was generated by 
```
cat_fastq {input.FQ} | minimap2 -t {threads} -K 500M -ax map-ont --MD -Y {input.IDX} - | samtools sort -@ {threads} -O BAM -o {output.BAM} - && samtools index -@ {threads} {output.BAM}
```
Nanoplot version==1.29.0
Thank you for your help.

**Comment by wdecoster** [comment]
Could you perhaps share your bam file for me to reproduce this?

**Comment by weir12** [comment]
@lpantano   @wdecoster 
Updating `statmodels`  package to version  **0.11.1** is useful for shooting my trouble.
Although `nanoplot` is based on an older version.
https://github.com/wdecoster/NanoPlot/blob/b9c6191dedb0773f58383aa8395cb438ffe3fe1a/setup.py#L44

If you find that strategy is NOT working, I would like to provide a demo of my bam file.
Thanks

**Comment by wdecoster** [comment]
Ah that's good to hear! Thanks for letting me know!

**Comment by wdecoster** [comment]
It's at least quite similar. The plot which causes this error uses seaborn code, which uses either statsmodels (if available) or scipy for the KDE calculation. The error is not raised with scipy, it seems.

**Comment by emollier** [comment]
I confirm your observation, the scipy codepath works well.  I'm a bit concerned about what happens when statsmodels ends up pulled on the system of NanoPlot users for whatever reason, but the core issue seems in the hands of statsmodels developers now.
Thank you for having spotted that option!

**Comment by emollier** [comment]
For the record, I tried [that patch on seaborn side](https://github.com/mwaskom/seaborn/commit/09fef026ad89a299e13db44fa5b92885fb5b2823) and the test base went through even though statsmodels was available on my machine.  The code just falls back to Scipy when statsmodel issues the runtime error.

The fix is brought in seaborn 0.10.1, so maybe solving this once and for all will just be a matter of bumping to that level hopefuly.

Have a nice day!  :)

**Comment by wdecoster** [comment]
Thanks so much for looking into this! I have changed the minimal required seaborn version to be 0.10.1, so let's hope we won't see this error again :)
